Only invalidate session if authenticated

wireframe opened the pull request:
Fix #721

marcoow commented on addon/mixins/data-adapter-mixin.js:
```diff
@@ -94,8 +95,10 @@ export default Ember.Mixin.create({
+      if (this.get('session.isAuthenticated')) {
+        this.get('session').invalidate();
+      }
+      return new DS.InvalidError(['Session is not authenticated']);
```
I don't think this should return an `InvalidError` as a 401 response doesn't indicate that the record is invalid but that the request wasn't allowed due to missing authorization or insufficient rights. When returning an `InvalidError` Ember Data would actually try to get errors for the record from the error which isn't really applicable in this case as the error doesn't apply to the record but to the request as such.

wireframe replied:
ah, ok.  I can revert that then, and can be addressed as a separate branch if needed.

The error in my case (#721) is related to session invalidation anyways.

marcoow commented:
Would be good to have a test for this. Also this breaks existing tests.

wireframe commented:
thanks for the feedback @marcoow.

I've added unit tests and rebased against latest master to cleanup the history.  let me know if there's anything else you'd like to see done.

marcoow commented:
thanks!
